src/lib/autoinstall/clients/inst_autoinit.rb

```diff
@@ -0,0 +1,368 @@
+            Yast::Packages.Initialize_StageInitial(show_popup, base_url, log_url, product.dir)
+            # select the product to install
+            Yast::Pkg.ResolvableInstall(product.details.product, :product, "")
+            # initialize addons and the workflow manager
+            Yast::AddOnProduct.SetBaseProductURL(base_url)
+            Yast::WorkflowManager.SetBaseWorkflow(false)
+            Yast::AutoinstFunctions.reset_product
+          # report error only for installation or if autoupgrade contain addon with relative url.
+          # This way autoupgrade for Full medium on registered system
+          # can autoupgrade with empty profile.
+          elsif !Yast::Mode.autoupgrade || addons_relurl
+            if Yast::Mode.autoupgrade && addons_relurl
+              msg = _("None or wrong base product has been defined " \
+                "in the AutoYaST configuration file. " \
+                "It needs to be specified as base for addons that use relurl scheme." \
+                "Please check the <b>products</b> entry in the <b>software</b> section.<br><br>" \
+                "Following base products are available:<br>")
+            else
+              msg = _("None or wrong base product has been defined " \
+                "in the AutoYaST configuration file. " \
+                "Please check the <b>products</b> entry in the <b>software</b> section.<br><br>" \
+                "Following base products are available:<br>")
+            end
+            Yast::AutoinstFunctions.available_base_products.each do |product|
+              # FIXME: here we abuse knowledge that base product is ProductLocation and not Product
+              msg += "#{product.details.product} (#{product.details.summary})<br>"
+            end
+            Yast::Popup.LongError(msg) # No timeout because we are stopping the installation/upgrade.
+            return :abort
+          end
+
+
+        end
+
+        if !(Yast::Mode.autoupgrade && Yast::AutoinstConfig.ProfileInRootPart)
+          @ret = Yast::WFM.CallFunction("inst_system_analysis", [])
+          return @ret if @ret == :abort
+        end
+
+        if Yast::Builtins.haskey(Yast::Profile.current, "iscsi-client")
+          Yast::Builtins.y2milestone("iscsi-client found")
+          Yast::WFM.CallFunction(
+            "iscsi-client_auto",
+            ["Import", Yast::Ops.get_map(Yast::Profile.current, "iscsi-client", {})]
+          )
+          Yast::WFM.CallFunction("iscsi-client_auto", ["Write"])
+        end
+
+        if Yast::Builtins.haskey(Yast::Profile.current, "fcoe-client")
+          Yast::Builtins.y2milestone("fcoe-client found")
+          Yast::WFM.CallFunction(
+            "fcoe-client_auto",
+            ["Import", Yast::Ops.get_map(Profile.current, "fcoe-client", {})]
+          )
+          Yast::WFM.CallFunction("fcoe-client_auto", ["Write"])
+        end
+
+        if !Y2Packager::MediumType.offline? && !Yast::AutoinstFunctions.selected_product &&
```

Comment:
Three negative conditions in a row. What about `Y2Packager::MediumType.online? && !Yast::AutoinstFunctions.selected_product && Yast::Mode.autoinst`? Or simply using the [De Morgan's Law](https://en.wikipedia.org/wiki/De_Morgan's_laws): `!(Y2Packager::MediumType.offline || Yast::AutoinstFunctions.selected_product || Yast::Mode.autoupgrade)`?

You know what? Just use what is more readable to you.

Reply:
sadly first part (online instead of offline) cannot be used due to openSUSE, even I am not sure how many people use autoupgrade on opensuse. The second form looks better for me and maybe in the end I will find even better.